Improve performance of array integer indexing

lgeiger opened the pull request:
This PR tries to improve performance of array integer indexing as a first step towards addressing #2878.

Since it seems like using `jit` as suggested in #2891 is difficult I've limited this PR to the case of indexing using a single Python integer outside of `jit` to keep the diff somewhat manageable. Some of the changes should also benefit other cases.

I split the PR into multiple commits that are best reviewed one by one. Also please let me know if these changes would prevent future planned optimisation that I am not aware of. If so I'm happy to close this, or split out some of the optimisation into separate PR's.

## Changes
- 09134b36bd0a3a6fc10609b8a7a28d93c69b2e93 prevents the conversion of Python integers to DeviceArrays and uses plain Python code in favour of `lax` operations for index normalisation when possible. This moves the scalar index computations to the CPU instead of running on an accelerator. I currently can't easily measure the performance impact of this but for me this seems to be preferred for statically known indices.
- 75a092e027b26bcecf832c8cbfe088e98ec8c08d prevents unnecessary calls to `lax.expand_dims`
- b8d70d31a5187c3c4cf72743b3f96a9026627904 removes explicit conversion using `asarray`
- cf8e8ed46062fce095ce76b001626a4d3c248a3f and 19ac51e10339c465f7c40ea3b9acb966320c07a0 remove expensive calls to `core.get_aval` to speedup checks

## Latency impact

I've measured indexing performance using a simple toy example on CPU:
```python
from jax import numpy as jnp

a = jnp.arange(4000)
a[5]
%timeit a[6].block_until_ready()
```

This PR improves performance by a factor of **4.8x**. Jax is still a lot slower than plain numpy but this already seems like a decent improvement.

| commit                                   |   time | reduction |
| ---------------------------------------- | -----: | --------: |
| 88ac4edf63625db10f92d5368a598205aac1deb4 (main baseline) | 468 µs |           |
| 09134b36bd0a3a6fc10609b8a7a28d93c69b2e93 | 225 µs | - 52 %    |
| 75a092e027b26bcecf832c8cbfe088e98ec8c08d | 168 µs | - 25 %    |
| b8d70d31a5187c3c4cf72743b3f96a9026627904 | 143 µs | - 15 %    |
| cf8e8ed46062fce095ce76b001626a4d3c248a3f | 125 µs | - 13 %    |
| 19ac51e10339c465f7c40ea3b9acb966320c07a0 |  98 µs | - 22 %    |

lgeiger committed:
Speedup index computations for integer indexing

lgeiger committed:
Prevent unnecessary `expand_dims`

lgeiger committed:
Remove unnecessary `asarray` call

lgeiger committed:
Do not call `concrete_aval` for basic integer index checks

lgeiger committed:
Speedup `_expand_bool_indices` when passing basic integer indices

lgeiger commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3417,15 +3417,20 @@ def _take(a, indices, axis: Optional[int] = None, out=None, mode=None):
-    lax.lt(index, _lax_const(index, 0)),
-    lax.add(index, axis_size_val),
-    index)
+  try:
+    idx = core.concrete_or_error(operator.index, index, "normalize index")
+    ndim = core.concrete_or_error(operator.index, axis_size, "normalize axis_size")
+    return _canonicalize_axis(idx, ndim)
+  except (TypeError, ValueError, IndexError, core.ConcretizationTypeError):
```
Not sure if this is the best way to check for scalar Python indices but it seems quite readable and is fast.
- `operator.index` might throw a `TypeError `
- `_canonicalize_axis` might throw a `ValueError`
- `core.concrete_or_error` might throw a `ConcretizationTypeError `
-  During construction of a `ConcretizationTypeError` an `IndexError` might be thrown similar to what was reported in #7426

YouJiacheng replied:
Use `canonicalize_axis` for normalize index may cause confusing error:
`ValueError(f"axis {axis} is out of bounds for array of dimension {num_dims}")`

jakevdp commented:
Related: #10150

lgeiger commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3519,7 +3524,6 @@ def _rewriting_take(arr, idx, indices_are_sorted=False, unique_indices=False,
   # Computes arr[idx].
   # All supported cases of indexing can be implemented as an XLA gather,
   # followed by an optional reverse and broadcast_in_dim.
-  arr = asarray(arr)
```
`_rewriting_take` is only called from within `__getitem__` so I think it is fine to remove here. Let me know if `asarray` is still needed for some other reason here.

jakevdp replied:
Seems reasonable - I can't think of any reason why the argument would not already be an array

lgeiger commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3866,6 +3882,8 @@ def _eliminate_deprecated_list_indexing(idx):
   # non-tuple sequence containing slice objects, [Ellipses, or newaxis
   # objects]". Detects this and raises a TypeError.
   if not isinstance(idx, tuple):
+    if isinstance(idx, (int, slice)):
+      return (idx,)
```
This is a tiny improvement to early exit in case of integer indices. `isinstance(idx, typing.Sequence)` is quite slow since it calls `__instancecheck__`. `isinstance(idx, collections.abc.Sequence)` would be faster in Python 3.9, but early exiting for the simple case shouldn't complicate things too much here.

YouJiacheng commented:
Just curious, does this PR affect performance after JIT?

lgeiger committed:
Add jaxpr test to ensure that no normalization happens for constant indices

lgeiger commented:
> Related: #10150

@jakevdp Thanks for mentioning that PR. I rebased and added a small test in 8054f426e13346d8c09022704761824651ef983c to ensure that only a gather operation is present when using constant integer indices.

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3417,17 +3417,22 @@ def _take(a, indices, axis: Optional[int] = None, out=None, mode=None):
+    return _canonicalize_axis(idx, ndim)
+  except (TypeError, ValueError, IndexError, core.ConcretizationTypeError):
+    if issubdtype(_dtype(index), np.unsignedinteger):
+      return index
```
This block can go before the `try` statement I think

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3417,17 +3417,22 @@ def _take(a, indices, axis: Optional[int] = None, out=None, mode=None):
-    index)
+  try:
+    idx = core.concrete_or_error(operator.index, index, "normalize index")
+    ndim = core.concrete_or_error(operator.index, axis_size, "normalize axis_size")
```
I think you can simplify this: concrete tracers implement the `__index__` method, so if the concretization error is not going to be user-facing, there's no benefit to calling `core.concrete_or_error`. Instead you could just call `operator.index` directly, and you'd only have to catch the `TypeError`.

jakevdp replied:
Actually, I just realized that if `operator.index` is called on a non-concrete tracer, it will result in a `TracerIntegerConversionError`, so you'll have to also catch that.

lgeiger replied:
Great, this is much nicer! `TracerIntegerConversionError` is a subclass of `TypeError`, so this will already be caught together with errors from `operator.index`.
I changed it in the latest commit.

YouJiacheng replied:
Hi! - @jakevdp , concrete tracer will raise exception on `__index__` but not `int`!
```python
from jax import grad
grad(lambda x: print(x.aval) or print(int(x)) or print(x.__index__()) or x)(1.0)
# ConcreteArray(1.0, dtype=float32, weak_type=True)
# 1
# jax._src.errors.TracerIntegerConversionError: The __index__() method was called on the JAX Tracer object 
# Traced<ConcreteArray(1.0, dtype=float32, weak_type=True)>with<JVPTrace(level=2/0)
```

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3417,17 +3417,22 @@ def _take(a, indices, axis: Optional[int] = None, out=None, mode=None):
+    if issubdtype(_dtype(index), np.unsignedinteger):
+      return index
+    if core.is_constant_dim(axis_size):
+      axis_size_val = _lax_const(index, axis_size)
```
The rest of this should probably not be in the `except` block, because it will cause confusing tracebacks if something in here raises an error. perhaps use `except: pass` followed by the rest?

lgeiger replied:
👍  Changed it. This reduces the diff of the PR.

lgeiger committed:
Simplify `_normalize_index`

lgeiger commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3417,6 +3417,11 @@ def _take(a, indices, axis: Optional[int] = None, out=None, mode=None):
+  try:
+    idx, ndim = operator.index(index), operator.index(axis_size)
+    return _canonicalize_axis(idx, ndim)
+  except (TypeError, ValueError):
```
Not catching `ValueError`s here would bring the behaviour closer to numpy but would [change the current behaviour for out of bound indexing](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#out-of-bounds-indexing) resulting in inconsistent behaviour.

jakevdp replied:
In that case maybe it's cleaner to not call canonicalize axis, and rather do this:
```python
return idx + ndim if idx < 0 else idx
```
then we don't have to worry about the `ValueError`

jakevdp replied:
Also, I think a more idiomatic way to write this may be
```python
try:
  idx, ndim = operator.index(index), operator.index(axis_size)
except TypeError:
  pass
else:
  return idx + ndim if idx < 0 else idx
```
It makes clear where the TypeError is expected, and avoids catching unexpected errors.

lgeiger replied:
That's a lot cleaner. Done 👍

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3897,13 +3915,11 @@ def _eliminate_deprecated_list_indexing(idx):
-          or isinstance(i, list) and i and _all(_is_scalar(e)
-          and issubdtype(_dtype(e), np.bool_) for e in i))
+  if isinstance(i, core.Tracer):
+    i = i.aval
```
Why did you switch to an instance check here but continue using `try: core.get_aval()` in other places?

lgeiger replied:
`core.get_aval()` will always create a `DeviceArray` if the argument is not a `Tracer`. This is unnecessary when calling it with a scalar Python index and is also quite a bit slower than using an `isinstance` check:
```python
In [3]: %timeit core.get_aval(3)
8.94 µs ± 260 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit isinstance(3, core.Tracer)
67.2 ns ± 0.969 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Since these checks happen quite often this was noticeable in the performance profile, so I tried to remove it where needed. I left calls to `core.get_aval` untouched in all other code paths to keep this PR limited to indexing with a single Python indices, but I'm happy take a look at optimising other forms of constant indexing in a future PR if that would be appreciated.

jakevdp commented on tests/lax_numpy_indexing_test.py:
```diff
@@ -871,6 +871,15 @@ def testTrivialGatherIsntGenerated(self):
+    x = jnp.arange(10)
+    jaxpr = jax.make_jaxpr(lambda: x[idx])()
+    self.assertEqual(len(jaxpr.jaxpr.eqns), 1)
+    self.assertEqual(jaxpr.jaxpr.eqns[0].primitive, lax.gather_p)
```
I might add another test here to confirm that if the index is a static array with `ndim > 0`, that we use traced operations rather than treating the index array as a constant. From your code I don't think there's any chance that would slip through, but it would be useful if there are any future refactorings.

lgeiger replied:
I think that's already covered in the following test, isn't it?
https://github.com/google/jax/blob/0ed29b63f08fa404b1e837b4ab8b96b481c4b78f/tests/lax_numpy_indexing_test.py#L623-L634

jakevdp replied:
Ah, I missed that, thanks

jakevdp reviewed (changes requested):
Thanks! Looks good, a couple more comments

jakevdp reviewed (approved):
Looks great, thanks!

jakevdp commented on tests/lax_numpy_indexing_test.py:
```diff
@@ -871,6 +871,15 @@ def testTrivialGatherIsntGenerated(self):
 
+  @parameterized.named_parameters(
+      {"testcase_name": f"_{idx}", "idx": idx}
+      for idx in (-3, 0, 5, np.array(6), jnp.array(7)))
```
Creating a `jnp.array` here at the global scope is not supported in some of our test frameworks, because the backend has not yet been initialized. I'd suggest doing something like
```
for idx in (-3, 0, 5)
for idx_type in (int, np.array, jnp.ndarray)
```
and then convert idx as necessary.

lgeiger replied:
:+1: changed it.

lgeiger commented on tests/lax_numpy_indexing_test.py:
```diff
@@ -871,6 +871,18 @@ def testTrivialGatherIsntGenerated(self):
+      {"testcase_name": f"_{idx_type_name}_{idx}", "idx": idx, "idx_type": idx_type}
+      for idx in (-3, 0, 5)
+      for idx_type_name, idx_type in (
+        ("int", int), ("np.array", np.array), ("jnp.array", jnp.array)))
```
Let me know if there is a better way to get the `idx_type` name automatically for the test name.

lgeiger committed:
Expand constant indexing test to check `slice`

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3647,8 +3655,17 @@ def _merge_static_and_dynamic_indices(treedef, static_idx, dynamic_idx):
+def _is_basic_int_index(x):
+  if isinstance(x, core.Tracer):
+    aval = x.aval
+    return (isinstance(aval, (ConcreteArray, ShapedArray)) and
```
`ConcreteArray` is a subclass of `ShapedArray`, so `isinstance(aval, ShapedArray)` should be sufficient.

lgeiger replied:
Updated in bb518728a576a67f3e4ff9c48552dd58600b8df7

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -3647,8 +3655,17 @@ def _merge_static_and_dynamic_indices(treedef, static_idx, dynamic_idx):
+  except TypeError:
+    return False
+  else:
+    return not isinstance(x, bool)
```
I think the logic here misses one corner case, though the behavior is deprecated in the current version of numpy:
```python
>>> idx = np.array([True], dtype=bool)[0]
>>> operator.index(idx)
DeprecationWarning: In future, it will be an error for 'np.bool_' scalars to be interpreted as an index
1
```

lgeiger replied:
Good catch, I wasn't aware of that. f23efb9f50880fa479323d35b8616a94e8fce37c should handle this case correctly.

lgeiger committed:
Simplify isinstance check

lgeiger committed:
Fix check for numpy bool indices

jakevdp commented:
It looks like this breaks some jax2tf tests, namely `jax/experimental/jax2tf/tests/shape_poly_test.py:ShapePolyPrimitivesTest.test_prim_getitem_op=poly_idx=slice-poly_noxla_poly_axes=[0]` (this test is not run as part of the github CI)

lgeiger commented:
For now I reverted 8e501651d54e765b4b65907e7612015be5e0263e which should fix the broken test.

jakevdp commented:
Thanks for the updates.

So, overall my concern here is that this PR touches a *lot* of parts of the code, all of which have potential side-effects that are difficult for me to predict. The passing tests are a good sign, but I'm not confident that our test coverage is complete when it comes to things like all the various ways that arrays can be indexed. A PR like this seems ripe for unanticipated and unintended downstream consequences.

Unfortunately, I'm not really sure what to suggest – it's hard to check for unknown issues. I'll pull-in the changes and run a larger suite of internal tests in case that shows us something.

jakevdp commented:
Well, additional testing didn't find any failures. I think we're in good shape

lgeiger commented:
Thanks for the detailed review and the help!
In case any unforeseen problems arise, I'm happy to take another look or extract  10addfd040acbbd1972a6697bf9c8134a1248043 and 38cee9f68644cb27307513ad56f2dad9134b5259 into separate PRs.
